### PR TITLE
Implement `listIndexes` for SQLite

### LINKED ISSUE #3175: Implement `listIndexes` for SQLite
### What should be done?

TBD

### PR DESCRIPTION
**Pull request by rumyantseva**
# Description

Closes #3175.

## Readiness checklist

- [ ] I added/updated unit tests (and they pass).
- [x] I added/updated integration/compatibility tests (and they pass).
- [x] I added/updated comments and checked rendering.
- [x] I made spot refactorings.
- [ ] I updated user documentation.
- [ ] I ran `task all`, and it passed.
- [x] I ensured that PR title is good enough for the changelog.
- [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Milestone (`Next`), Labels, Project and project's Sprint fields.
- [x] I marked all done items in this checklist.

### CODE REVIEW

File: internal/backends/collection.go [hunk 1]
```diff
@@ -224,6 +226,42 @@ func (cc *collectionContract) Stats(ctx context.Context, params *CollectionStats
...
+
+// IndexKeyPair consists of a field name and a sort order that are part of the index.
+type IndexKeyPair struct {
+	Field string
+	Order types.SortType
+}
+
+// ListIndexes returns information about indexes in the database.
+//
+// Database or collection may not exist; that's not an error.
```
Comment (chilagrow) [inline internal/backends/collection.go:255]: I think non existent database and collection return error for indexes :thinking:

File: internal/handlers/sqlite/msg_listindexes.go [hunk 2]
```diff
@@ -16,13 +16,96 @@ package sqlite
...
+
+	for _, index := range res.Indexes {
+		indexKey := must.NotFail(types.NewDocument())
+
+		for _, key := range index.Key {
+			indexKey.Set(key.Field, int32(key.Order))
+		}
+
+		indexDoc := must.NotFail(types.NewDocument(
+			"v", int32(2),
```
Comment (chilagrow) [inline internal/handlers/sqlite/msg_listindexes.go:85]: Took me a while to figure out `v: 2` is the version of index. I was thinking it was a return value for default index :see_no_evil:

A comment would be great here :hugs:

**Review by chilagrow — COMMENTED** [review]
Looks good, just a few comments

File: internal/backends/collection.go (hunk 1 above)
Comment (rumyantseva) [reply internal/backends/collection.go:255]: Yes, you are right! Somehow I lost that moment. I will add a test (and fix the logic).

File: internal/handlers/sqlite/msg_listindexes.go (hunk 2 above)
Comment (rumyantseva) [reply internal/handlers/sqlite/msg_listindexes.go]: Fixed

File: internal/backends/sqlite/collection.go [hunk 3]
```diff
@@ -287,6 +288,42 @@ func (c *collection) Stats(ctx context.Context, params *backends.CollectionStats
...
+	// TODO https://github.com/FerretDB/FerretDB/issues/3176
+	return &backends.ListIndexesResult{
+		Indexes: []backends.IndexInfo{
+			{
+				Unique: pointer.ToBool(true),
+				Name:   "_id_",
+				Key: []backends.IndexKeyPair{
+					{
+						Field: "_id",
+						Order: 1,
```
Comment (AlekSi) [inline internal/backends/sqlite/collection.go:319]: So are we using 1/-1 or types.Ascending/Descending?

**Review by chilagrow — COMMENTED** [review]
Let's apply @AlekSi's comment, otherwise looks good

File: internal/backends/collection.go [hunk 4]
```diff
@@ -224,6 +226,43 @@ func (cc *collectionContract) Stats(ctx context.Context, params *CollectionStats
...
+type ListIndexesParams struct{}
+
+// ListIndexesResult represents the results of Database.ListIndexesResult method.
+type ListIndexesResult struct {
+	Indexes []IndexInfo
+}
+
+// IndexInfo represents information about a single index.
+type IndexInfo struct {
+	Unique *bool
```
[the remark this answers is not in the document]
Comment (rumyantseva) [reply internal/backends/collection.go:239]: Not really. But let's keep it just `bool` for now, we can add a pointer later when we meet a particular case.

File: internal/handlers/sqlite/msg_listindexes.go [hunk 5]
```diff
@@ -16,13 +16,102 @@ package sqlite
...
+		return nil, lazyerrors.Error(err)
+	}
+
+	firstBatch := types.MakeArray(len(res.Indexes))
+
+	for _, index := range res.Indexes {
+		indexKey := must.NotFail(types.NewDocument())
+
+		for _, key := range index.Key {
+			indexKey.Set(key.Field, int32(key.Order))
```
[the remark this answers is not in the document]
Comment (rumyantseva) [reply internal/handlers/sqlite/msg_listindexes.go:86]: fixed

File: internal/backends/sqlite/collection.go (hunk 3 above)
Comment (rumyantseva) [reply internal/backends/sqlite/collection.go]: fixed

File: internal/handlers/sqlite/msg_listindexes.go [hunk 6]
```diff
@@ -16,13 +16,102 @@ package sqlite
...
+		if backends.ErrorCodeIs(err, backends.ErrorCodeCollectionNameIsInvalid) {
+			msg := fmt.Sprintf("Invalid collection name: %s", collection)
+			return nil, commonerrors.NewCommandErrorMsgWithArgument(commonerrors.ErrInvalidNamespace, msg, document.Command())
+		}
+
+		return nil, lazyerrors.Error(err)
+	}
+
+	res, err := c.ListIndexes(ctx, nil)
+	if backends.ErrorCodeIs(err, backends.ErrorCodeDatabaseDoesNotExist) ||
```
[the remark this answers is not in the document]
Comment (rumyantseva) [reply internal/handlers/sqlite/msg_listindexes.go]: fixed

File: internal/backends/collection.go [hunk 7]
```diff
@@ -224,6 +226,43 @@ func (cc *collectionContract) Stats(ctx context.Context, params *CollectionStats
...
+// IndexKeyPair consists of a field name and a sort order that are part of the index.
+type IndexKeyPair struct {
+	Field string
+	Order types.SortType
+}
+
+// ListIndexes returns information about indexes in the database.
+//
+// If database does not exist it returns ErrorCodeDatabaseDoesNotExist.
+// If collection does not exist it returns ErrorCodeCollectionDoesNotExist.
```
[the remark this answers is not in the document]
Comment (rumyantseva) [reply internal/backends/collection.go]: fixed

File: internal/backends/collection.go [hunk 8]
```diff
@@ -224,6 +226,43 @@ func (cc *collectionContract) Stats(ctx context.Context, params *CollectionStats
...
+// ListIndexesResult represents the results of Database.ListIndexesResult method.
+type ListIndexesResult struct {
+	Indexes []IndexInfo
+}
+
+// IndexInfo represents information about a single index.
+type IndexInfo struct {
+	Unique *bool
+	Name   string
+	Key    IndexKey
```
[the remark this answers is not in the document]
Comment (rumyantseva) [reply internal/backends/collection.go]: fixed

File: internal/backends/collection.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (rumyantseva) [reply internal/backends/collection.go]: fixed

**Review by chilagrow — APPROVED** [review]
:rocket:

**Review by chilagrow — DISMISSED** [review]

File: internal/backends/collection.go [hunk 9]
```diff
@@ -224,6 +226,50 @@ func (cc *collectionContract) Stats(ctx context.Context, params *CollectionStats
...
+type IndexInfo struct {
+	Name   string
+	Key    []IndexKeyPair
+	Unique bool
+}
+
+// IndexKeyPair consists of a field name and a sort order that are part of the index.
+type IndexKeyPair struct {
+	Field string
+	Order IndexOrder
```
Comment (AlekSi) [inline internal/backends/collection.go:247]:
```suggestion
	Descending bool
```

To avoid any confusion

File: internal/backends/collection.go [hunk 10]
```diff
@@ -224,6 +226,50 @@ func (cc *collectionContract) Stats(ctx context.Context, params *CollectionStats
...
+	// IndexOrderAsc represents ascending sort order.
+	IndexOrderAsc IndexOrder = 1
+
+	// IndexOrderDesc represents descending sort order.
+	IndexOrderDesc IndexOrder = -1
+)
+
+// ListIndexes returns information about indexes in the database.
+//
+// If database or collection does not exist it returns ErrorCodeDatabaseDoesNotExist.
```
Comment (AlekSi) [inline internal/backends/collection.go:263]: https://github.com/FerretDB/FerretDB/blob/ed54533eef28f969df5eb9b6d2e8e2c8936c6848/internal/backends/doc.go#L35-L36

File: internal/backends/collection.go [hunk 11]
```diff
@@ -224,6 +226,50 @@ func (cc *collectionContract) Stats(ctx context.Context, params *CollectionStats
...
+func (cc *collectionContract) ListIndexes(ctx context.Context, params *ListIndexesParams) (*ListIndexesResult, error) {
+	defer observability.FuncCall(ctx)()
+
+	res, err := cc.c.ListIndexes(ctx, params)
+	checkError(err, ErrorCodeDatabaseDoesNotExist)
```
Comment (AlekSi) [inline internal/backends/collection.go:268]: That should be `ErrorCodeCollectionDoesNotExist` to match `DropCollection` and `RenameCollection`

File: internal/backends/collection.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (AlekSi) [reply internal/backends/collection.go]: Fixed by returning the different error from those methods. Why?

File: internal/backends/sqlite/collection.go [hunk 12]
```diff
@@ -287,6 +287,44 @@ func (c *collection) Stats(ctx context.Context, params *backends.CollectionStats
...
+	if db == nil {
+		return nil, backends.NewError(
+			backends.ErrorCodeDatabaseDoesNotExist,
+			lazyerrors.Errorf("no ns %s.%s", c.dbName, c.name),
+		)
+	}
+
+	coll := c.r.CollectionGet(ctx, c.dbName, c.name)
+	if coll == nil {
+		// It enough to return ErrorCodeDatabaseDoesNotExist:
```
[the remark this answers is not in the document]
Comment (rumyantseva) [reply internal/backends/sqlite/collection.go:302]: The comment is left to document the reason for this decision. One might think that it's a typo and `ErrorCodeCollectionDoesNotExist` must be returned here instead.

[the remark this answers is not in the document]
Comment (AlekSi) [reply internal/backends/sqlite/collection.go:302]: The decision isn't made there, in all implementations – it is made in the interface contract.

[the remark this answers is not in the document]
Comment (rumyantseva) [reply internal/backends/sqlite/collection.go:302]: Ok, then I'll move the comment there.

**Review by AlekSi — APPROVED** [review]

**Review by chilagrow — APPROVED** [review]
:rocket:

**Pull request merged**